mixer/adapter/circonus/circonus.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package circonus
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	cgm "github.com/circonus-labs/circonus-gometrics"
+	"github.com/circonus-labs/circonus-gometrics/checkmgr"
+	"istio.io/istio/mixer/adapter/circonus/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/metric"
+	"log"
+	"net/url"
+	"sync"
+	"time"
+)
+
+type (
+	builder struct {
+		adpCfg      *config.Params
+		metricTypes map[string]*metric.Type
+	}
+
+	handler struct {
+		cm            cgm.CirconusMetrics
+		env           adapter.Env
+		metrics       map[string]config.Params_MetricInfo_Type
+		lastFlushed   time.Time
+		flushLock     sync.Mutex
+		flushInterval time.Duration
+	}
+)
+
+// ensure types implement the requisite interfaces
+var _ metric.HandlerBuilder = &builder{}
+var _ metric.Handler = &handler{}
+
+// bridge stdlog to env.Logger()
+type logToEnvLogger struct {
+	env adapter.Env
+}
+
+// Build constructs a circonus-gometrics instance and sets up the handler
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+
+	// register cgm.Flush() with env.ScheduleWork() to wrap goroutines
+	var cm *cgm.CirconusMetrics
+	ec := make(chan error, 1)
+
+	env.ScheduleWork(
+		func() {
+
+			bridge := &logToEnvLogger{env: env}
+
+			cmc := &cgm.Config{
+				CheckManager: checkmgr.Config{
+					Check: checkmgr.CheckConfig{
+						SubmissionURL: b.adpCfg.SubmissionUrl,
+					},
+				},
+				Log:      log.New(bridge, "", 0),
+				Debug:    true, // enable [DEBUG] level logging for env.Logger
+				Interval: "0s", // do not autoflush
+			}
+
+			var err error
+			cm, err = cgm.NewCirconusMetrics(cmc)
+			ec <- err
+		})
+
+	select {
+	case err := <-ec:
+		if err != nil {
+			env.Logger().Errorf("Error instantiating cgm: %v", err)
+			return nil, err
+		}
+	}
+
+	// wait until cgm background tasks initialized under ScheduleWork have completed
+	for !cm.Ready() {
```

Comment:
Ick, there's no better way than polling?

Reply:
Yup - ick. I had to poll here, or I got an error because mixer tried to start up with cm uninitialized.